src/legacy/core_plugins/embeddable_api/public/actions/action.ts

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { EuiContextMenuItemIcon } from '@elastic/eui';
+
+import { IEmbeddable } from '../embeddables';
+
+export interface ActionContext<
+  TEmbeddable extends IEmbeddable = IEmbeddable,
+  TTriggerContext extends {} = {}
+> {
+  embeddable: TEmbeddable;
+  triggerContext?: TTriggerContext;
+}
+
+export abstract class Action<
+  TEmbeddable extends IEmbeddable = IEmbeddable,
+  TTriggerContext extends {} = {}
+> {
+  /**
+   * Determined the order when there is more than one action matched to a trigger.
+   * Higher numbers are displayed first.
+   */
+  public order: number = 0;
+  public abstract readonly type: string;
+
+  constructor(public readonly id: string) {}
+
+  /**
+   * Optional icon that can be displayed along with the title.
+   */
+  public getIcon(
+    context: ActionContext<TEmbeddable, TTriggerContext>
+  ): EuiContextMenuItemIcon | undefined {
+    return undefined;
+  }
+
+  /**
+   * Returns a title to be displayed to the user.
+   * @param context
+   */
+  public abstract getDisplayName(context: ActionContext<TEmbeddable, TTriggerContext>): string;
+
+  /**
+   * Returns a promise that resolves to true if this action is compatible given the context,
+   * otherwise resolves to false.
+   */
+  public async isCompatible(
+    context: ActionContext<TEmbeddable, TTriggerContext>
+  ): Promise<boolean> {
+    return true;
+  }
+
+  /**
+   * If this returns something other than undefined, this is used instead of execute when clicked.
+   */
+  public getHref(context: ActionContext<TEmbeddable, TTriggerContext>): string | undefined {
```

Comment:
Yea, this is because people want to be able to right click on certain menu items and say "open in new tab".   Good call on me not taking this into account in `executeTriggerActions`... fixed.  

I actually think we should perhaps always show the context menu except make a single exception for the `APPLY_FILTER_ACTION` action just to preserve existing functionality.  But we can discuss that more later.